### PR TITLE
More strict argument type-checking

### PR DESCRIPTION
**Pull request by ylemkimon**
- Remove `numArgs` and `numOptionalArgs`
- Separate `optionalArgTypes` from `argTypes`
- Make `argTypes` and `optionalArgTypes` explicit 
- Type handler `args` using `$TupleMap`

### CODE REVIEW

File: src/Parser.js [hunk 1]
```diff
@@ -441,6 +466,7 @@ export default class Parser {
      */
     callFunction(
         name: string,
+        // TODO(ylem): args should be $TupleMap<ARGTYPES, ArgNode>
         args: AnyParseNode[],
```
Comment (ylemkimon) [inline src/Parser.js]: TODO(1)

File: src/Parser.js [hunk 2]
```diff
@@ -463,48 +489,50 @@ export default class Parser {
...
      * Parses the arguments of a function or environment
      */
-    parseArguments(
+    parseArguments<ARGTYPES: ArgType[]>(
         func: string,   // Should look like "\name" or "\begin{name}".
-        funcData: FunctionSpec<*> | EnvSpec<*>,
+        funcData: FunctionSpec<*, ARGTYPES> | EnvSpec<*, ARGTYPES>,
     ): {
+        // TODO(ylem): args should be $TupleMap<ARGTYPES, ArgNode>
         args: AnyParseNode[],
```
Comment (ylemkimon) [inline src/Parser.js]: TODO(2)

File: src/Parser.js [hunk 3]
```diff
@@ -515,11 +543,12 @@ export default class Parser {
      */
     parseGroupOfType(
         name: string,
-        type: ?ArgType,
+        type: ArgType,
         optional: boolean,
         greediness: ?number,
         consumeSpaces: boolean,
     ): ?AnyParseNode {
+    // TODO(ylem): Return type should be ParseNode<$ElementType<ArgNodeType, type>>
```
Comment (ylemkimon) [inline src/Parser.js]: TODO(3)

File: src/defineEnvironment.js [hunk 4]
```diff
@@ -24,21 +24,22 @@ type EnvContext = {|
  *  - args: an array of arguments passed to \begin{name}
  *  - optArgs: an array of optional arguments passed to \begin{name}
  */
-type EnvHandler = (
+type EnvHandler<NODETYPE: NodeType, ARGTYPES: ArgType[]> = (
     context: EnvContext,
-    args: AnyParseNode[],
+    args: $TupleMap<ARGTYPES, ArgNode>,
+    // TODO(ylem): Type optArgs
     optArgs: (?AnyParseNode)[],
```
Comment (ylemkimon) [inline src/defineEnvironment.js]: TODO(4)

File: src/Parser.js [hunk 5]
```diff
@@ -441,6 +466,7 @@ export default class Parser {
      */
     callFunction(
         name: string,
+        // TODO(ylem): args should be $TupleMap<ARGTYPES, ArgNode>
```
Comment (kevinbarabash) [inline src/Parser.js:469]: Does `$TupleMap<ARGTYPES, ArgNode>` work in Flow?

**Comment by kevinbarabash** [comment]
This is interesting.  Sorry for not having looked at this until now. 😞  Maybe we can revisit this after the move to TS.

File: src/Parser.js (hunk 5 above)
Comment (ylemkimon) [reply src/Parser.js:469]: Unfortunately, I couldn't get it to work.

**Pull request closed without merging**